xapian-core/include/xapian/diversify.h

```diff
@@ -0,0 +1,133 @@
+/** @file diversify.h
+ *  @brief Diversification API
+ */
+/* Copyright (C) 2018 Uppinder Chugh
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#ifndef XAPIAN_INCLUDED_DIVERSIFY_H
+#define XAPIAN_INCLUDED_DIVERSIFY_H
+
+#if !defined XAPIAN_IN_XAPIAN_H && !defined XAPIAN_LIB_BUILD
+#error "Never use <xapian/diversify.h> directly; include <xapian.h> instead."
+#endif
+
+#include <xapian/attributes.h>
+#include <xapian/cluster.h>
+#include <xapian/mset.h>
+#include <xapian/types.h>
+#include <xapian/visibility.h>
+
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace Xapian {
+
+/** Class for diversifying an MSet using GLS-MPT as given
+ *  in the paper: Scalable and Efficient Web Search Result
+ *  Diversification Naini et al. 2016
+ */
+class XAPIAN_VISIBILITY_DEFAULT Diversify {
+    /// Top-k documents of given mset are diversified
+    unsigned int k;
+
+    /// MPT parameters
+    double lambda, b, sigma_sqr;
+
+    /// Store each document from given mset as a point and its relevance score
+    std::unordered_map<Xapian::docid, Xapian::Point> points;
+
+    /// Store the relevance score of each document
+    std::unordered_map<Xapian::docid, double> scores;
+
+    /// Store pairwise cosine similarities of documents of given mset
+    std::map<std::pair<Xapian::docid, Xapian::docid>, double> pairwise_sim;
+
+    /// Store docids of top k diversified documents
+    std::vector<Xapian::docid> main_dmset;
+
+    /** Initial diversified document set
+     *
+     *  Return top-k documents of mset as vector of Points, which
+     *  represent the initial diversified document set.
+     *
+     *  @param source	MSet object containing the documents of which
+     *			top-k are to be diversified
+     */
+    void initialise_points(const Xapian::MSet &source);
```

Comment:
The current code base uses rather a mixture of placements of the `&` for a reference type (and similarly for `*` for pointer types) - you'll probably see all of `const Xapian::MSet &source`, `const Xapian::MSet & source` and `const Xapian::MSet& source`.

However we're aiming to standardise on the latter as it seems to be the preferred modern C++ style - if you could try to follow that in new code that'd be great.

Reply:
I had wondered why there was such a mixture in the code base. Anyways, pushed a fixup commit for the modifications.

Reply:
The oldest code was written by Richard and myself back in 1999.  Then we were both fairly new to C++ and from a C background where it's common to write `char *p`.  One argument for that over the `char* p` form is that you can write:

```
char *p, q;
```

which is the same as:

```
char *p;
char q;
```

whereas:

```
char* p, q;
```

reads as if q is a `char*`.

Also:

```
char *p, *q;
```

reads OK, but where do you put the `*` for q with the modern C++ style?  About the best I can see is:

```
char* p, * q;
```

But such mixed declarations are more naturally split onto separate lines in C++ which encourages initialisation at the point of declaration.

At some point we switched to a space either side - not sure I really remember the logic as to why.

It would be good to do a mass clean-up at some point, but such things can be disruptive - it makes backporting patches harder, unmerged patches may stop applying cleanly, and it makes tools like `git blame` less useful (though with git you can use `-w` to ignore whitespace so it can see through this sort of change - we used to use subversion for version control and I don't think that allowed that).  Therefore we tend to hold off such things for a suitable time (such as shortly before cutting the next stable release branch, as then the older release branch will be moving to a very conservative backporting policy).